### PR TITLE
Unscented Kalman filter (UKF)

### PR DESCRIPTION
**Pull request by rolalaro**
- Created a class implementing the Unscented Kalman Filter
- Created 3 examples to show how to use the class

To do:

- [x] tutorial code
- [x] tutorial doc
- [x] doc: the equations of UKF
- [x] doc: links towards resources about UKF
- [x] examples: move current tutorial into examples

### CODE REVIEW

File: doc/tutorial/python/tutorial-install-python-bindings.dox [hunk 1]
```diff
@@ -672,26 +672,50 @@ If you have an issue that looks like:
...
-                  from /home/sfelton/.local/include/pybind11/pybind11.h:13,
-                  from /home/sfelton/software/visp_build/modules/python/bindings/src/robot.cpp:3:
+                  from /home/user/miniconda3/envs/wrapper3.9/include/python3.9/pyport.h:205,
+                  from /home/user/miniconda3/envs/wrapper3.9/include/python3.9/Python.h:50,
+                  from /home/user/.local/include/pybind11/detail/common.h:266,
+                  from /home/user/.local/include/pybind11/attr.h:13,
+                  from /home/user/.local/include/pybind11/detail/class.h:12,
+                  from /home/user/.local/include/pybind11/pybind11.h:13,
+                  from /home/user/software/visp_build/modules/python/bindings/src/robot.cpp:3:
   /usr/include/c++/11/type_traits: **In instantiation of ‘struct std::is_move_constructible<vpImage<double> >’:**
```
Comment (fspindle) [inline doc/tutorial/python/tutorial-install-python-bindings.dox:682]: To match the installation doc mentioning `$VISP_WS/visp` and `$VISP_WS/visp-build` instead of `/home/user/software/visp_build` I would rather use `/home/user/visp_ws/visp-build`

File: doc/tutorial/python/tutorial-install-python-bindings.dox [hunk 2]
```diff
@@ -672,26 +672,50 @@ If you have an issue that looks like:
...
   /usr/include/c++/11/type_traits:2209:11:   required by substitution of ‘template<class ... _Cond> using _Require = std::__enable_if_t<std::__and_< <template-parameter-1-1> >::value> [with _Cond = {std::__not_<std::__is_tuple_like<vpImage<double> > >, std::is_move_constructible<vpImage<double> >, std::is_move_assignable<vpImage<double> >}]’
   /usr/include/c++/11/bits/move.h:196:5:   required by substitution of ‘template<class _Tp> std::_Require<std::__not_<std::__is_tuple_like<_Tp> >, std::is_move_constructible<_Tp>, std::is_move_assignable<_Tp> > std::swap(_Tp&, _Tp&) [with _Tp = vpImage<double>]’
-  /home/sfelton/software/visp-sfelton/modules/core/include/visp3/core/vpImage.h:341:15:   required from ‘class vpImage<double>’
-  /home/sfelton/software/visp-sfelton/modules/core/include/visp3/core/vpImage.h:369:17:   required from here
+  /home/user/software/visp-user/modules/core/include/visp3/core/vpImage.h:341:15:   required from ‘class vpImage<double>’
+  /home/user/software/visp-user/modules/core/include/visp3/core/vpImage.h:369:17:   required from here
```
Comment (fspindle) [inline doc/tutorial/python/tutorial-install-python-bindings.dox:688]: Same here, instead of `/home/user/software/visp-user/` use `/home/user/visp_ws/visp/`

File: doc/tutorial/python/tutorial-install-python-bindings.dox [hunk 3]
```diff
@@ -672,26 +672,50 @@ If you have an issue that looks like:
...
+ 1432 |         : arg(base), value(reinterpret_steal<object>(detail::make_caster<T>::cast(
+      |                                                      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~^
+ 1433 |                          std::forward<T>(x), return_value_policy::automatic, {}))),
+      |                          ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+In file included from /home/user/miniconda3/envs/visp-python11/include/pybind11/cast.h:15,
+                 from /home/user/miniconda3/envs/visp-python11/include/pybind11/attr.h:14,
+                 from /home/user/miniconda3/envs/visp-python11/include/pybind11/detail/class.h:12,
+                 from /home/user/miniconda3/envs/visp-python11/include/pybind11/pybind11.h:13,
+                 from /home/user/visp_ws/visp_ukf/modules/python/bindings/src/core.cpp:3:
+\endverbatim
```
Comment (fspindle) [inline doc/tutorial/python/tutorial-install-python-bindings.dox:712]: Instead of `/home/user/visp_ws/visp_ukf/` use `/home/user/visp_ws/visp/`

File: modules/core/include/visp3/core/vpMatrix.h [hunk 4]
```diff
@@ -347,8 +347,15 @@ class VISP_EXPORT vpMatrix : public vpArray2D<double>
...
+  vpMatrix cholesky() const;
+#if defined(VISP_HAVE_EIGEN3)
+  vpMatrix choleskyByEigen3() const;
+#endif
+#if defined(VISP_HAVE_LAPACK)
+  vpMatrix choleskyByLapack() const;
+#endif
 
-  // Compute the exponential matrix of a square matrix
+    // Compute the exponential matrix of a square matrix
```
Comment (fspindle) [inline modules/core/include/visp3/core/vpMatrix.h:358]: Indentation problem

File: modules/core/src/math/matrix/vpMatrix.cpp [hunk 5]
```diff
@@ -128,7 +128,7 @@ void compute_pseudo_inverse(const vpMatrix &U, const vpColVector &sv, const vpMa
 
   unsigned int sv_size = sv.size();
   for (unsigned int i = 0; i < sv_size; ++i) {
-    if (sv[i] > (maxsv * svThreshold)) {
+    if (sv[i] >(maxsv * svThreshold)) {
```
Comment (fspindle) [inline modules/core/src/math/matrix/vpMatrix.cpp:131]: Should be `if (sv[i] > (maxsv * svThreshold)) {`. Is your visual code well configured ?

File: modules/core/src/math/matrix/vpMatrix.cpp [hunk 6]
```diff
@@ -4999,17 +4999,17 @@ vpColVector vpMatrix::getCol(unsigned int j) const { return getCol(j, 0, rowNum)
...
+[4, 4] =
+0  1  2  3
+4  5  6  7
+8  9 10 11
+12 13 14 15
+Row vector :
+4  5  6  7
+\endcode
+*/
 vpRowVector vpMatrix::getRow(unsigned int i) const { return getRow(i, 0, colNum); }
```
Comment (fspindle) [inline modules/core/src/math/matrix/vpMatrix.cpp:5013]: Revert these changes

File: modules/core/src/math/matrix/vpMatrix.cpp [hunk 7]
```diff
@@ -6088,7 +6088,7 @@ unsigned int vpMatrix::kernel(vpMatrix &kerAt, double svThreshold) const
 
   unsigned int rank = 0;
   for (unsigned int i = 0; i < nbcol; ++i) {
-    if (sv[i] > (maxsv * svThreshold)) {
+    if (sv[i] >(maxsv * svThreshold)) {
       rank++;
```
Comment (fspindle) [inline modules/core/src/math/matrix/vpMatrix.cpp:6092]: revert these changes

File: modules/core/src/math/matrix/vpMatrix.cpp [hunk 8]
```diff
@@ -6157,7 +6157,7 @@ unsigned int vpMatrix::nullSpace(vpMatrix &kerA, double svThreshold) const
 
   unsigned int rank = 0;
   for (unsigned int i = 0; i < nbcol; ++i) {
-    if (sv[i] > (maxsv * svThreshold)) {
+    if (sv[i] >(maxsv * svThreshold)) {
```
Comment (fspindle) [inline modules/core/src/math/matrix/vpMatrix.cpp:6160]: same here

File: modules/core/src/math/matrix/vpMatrix.cpp [hunk 9]
```diff
@@ -6234,7 +6234,7 @@ unsigned int vpMatrix::nullSpace(vpMatrix &kerA, int dim) const
   double maxsv = sv[0];
   unsigned int rank = 0;
   for (unsigned int i = 0; i < nbcol; ++i) {
-    if (sv[i] > (maxsv * 1e-6)) {
+    if (sv[i] >(maxsv * 1e-6)) {
```
Comment (fspindle) [inline modules/core/src/math/matrix/vpMatrix.cpp:6237]: same here

File: modules/core/src/math/matrix/vpMatrix.cpp [hunk 10]
```diff
@@ -6463,7 +6475,7 @@ double vpMatrix::cond(double svThreshold) const
   // Compute the rank of the matrix
   unsigned int rank = 0;
   for (unsigned int i = 0; i < nbcol; ++i) {
-    if (sv[i] > (maxsv * svThreshold)) {
+    if (sv[i] >(maxsv * svThreshold)) {
```
Comment (fspindle) [inline modules/core/src/math/matrix/vpMatrix.cpp:6478]: revert these changes

File: modules/core/src/math/matrix/vpMatrix_cholesky.cpp [hunk 11]
```diff
@@ -117,8 +117,8 @@ vpMatrix vpMatrix::inverseByCholesky() const
   return inverseByCholeskyOpenCV();
 #else
   throw(vpException(vpException::fatalError, "Cannot inverse matrix by "
-                                             "Cholesky. Install Lapack or "
-                                             "OpenCV 3rd party"));
+                    "Cholesky. Install Lapack or "
+                    "OpenCV 3rd party"));
```
Comment (fspindle) [inline modules/core/src/math/matrix/vpMatrix_cholesky.cpp:121]: replace with `throw(vpException(vpException::fatalError, "Cannot inverse matrix by Cholesky. Install Lapack or OpenCV 3rd party"));`

File: modules/core/test/math/testMatrixCholesky.cpp [hunk 12]
```diff
@@ -0,0 +1,197 @@
...
+  }
+  else {
+    std::cout << "Test failed" << std::endl;
+    return EXIT_FAILURE;
+  }
+#else
+  std::cout << "Test ignored: install Lapack or Eigen3" << std::endl;
+  return EXIT_SUCCESS;
+#endif
+}
```
Comment (fspindle) [inline modules/core/test/math/testMatrixCholesky.cpp:197]: I suggest to use `catch2` for new tests. See for example `testQuaternion.cpp`

File: tutorial/kalman/CMakeLists.txt [hunk 13]
```diff
@@ -0,0 +1,18 @@
+cmake_minimum_required(VERSION 3.5)
```
Comment (fspindle) [inline tutorial/kalman/CMakeLists.txt:1]: Copyright header is missing

File: tutorial/kalman/tutorial-ukf-linear-example.cpp [hunk 14]
```diff
@@ -0,0 +1,229 @@
...
+  point[0] = chi[0];
+  point[1] = chi[2];
+  return point;
+}
+
+int main(/*const int argc, const char *argv[]*/)
+{
+  const double dt = 1.; // Period of 1s
+  const double gt_dx = 1.; // Ground truth displacement along x axis between two measurements
+  const double gt_dy = 0.5; // Ground truth displacement along x axis between two measurements
```
Comment (fspindle) [inline tutorial/kalman/tutorial-ukf-linear-example.cpp:105]: Values are not realistic for a robotics system like a mobile robot. Instead of 1 sec, I suggest 0.010 sec, instead of 1. meter, I suggest 0.01 meter displacement along x and 0.005 meter along y

File: tutorial/kalman/tutorial-ukf-nonlinear-complex-example.cpp [hunk 15]
```diff
@@ -0,0 +1,633 @@
+/****************************************************************************
+ *
+ * ViSP, open source Visual Servoing Platform software.
+ * Copyright (C) 2005 - 2023 by Inria. All rights reserved.
```
Comment (fspindle) [inline tutorial/kalman/tutorial-ukf-nonlinear-complex-example.cpp:4]: 2005 - 2024

Comment (fspindle) [reply tutorial/kalman/tutorial-ukf-nonlinear-complex-example.cpp:4]: Check that in all new headers ou have `2005 - 2024` in the copyright header

File: tutorial/kalman/tutorial-ukf-nonlinear-complex-example.cpp [hunk 16]
```diff
@@ -0,0 +1,633 @@
...
+  const unsigned int nbLandmarks = sizeMeasurement / 2;
+  vpColVector mean(sizeMeasurement, 0.);
+  std::vector<double> sumCos(nbLandmarks, 0.);
+  std::vector<double> sumSin(nbLandmarks, 0.);
+  for (unsigned int i = 0; i < nbPoints; ++i) {
+    for (unsigned int j = 0; j < nbLandmarks; ++j) {
+      mean[2*j] += wm[i] * measurements[i][2*j];
+      sumCos[j] += wm[i] * std::cos(measurements[i][2*j+1]);
+      sumSin[j] += wm[i] * std::sin(measurements[i][2*j+1]);
+    }
```
Comment (fspindle) [inline tutorial/kalman/tutorial-ukf-nonlinear-complex-example.cpp:125]: With respect to sonaqube quality rules add parenthesis
```
      sumCos[j] += wm[i] * std::cos((measurements[i][2*j) + 1]);
      sumSin[j] += wm[i] * std::sin((measurements[i][2*j) + 1]);
```

File: tutorial/kalman/tutorial-ukf-nonlinear-example.cpp [hunk 17]
```diff
@@ -0,0 +1,339 @@
+/****************************************************************************
+ *
+ * ViSP, open source Visual Servoing Platform software.
+ * Copyright (C) 2005 - 2023 by Inria. All rights reserved.
+ *
```
Comment (fspindle) [inline tutorial/kalman/tutorial-ukf-nonlinear-example.cpp:5]: 2005 - 2024

File: tutorial/kalman/tutorial-ukf-nonlinear-example.cpp [hunk 18]
```diff
@@ -0,0 +1,339 @@
...
+  double m_x; // The position on the ground of the radar
+  double m_y; // The altitude of the radar
+  vpGaussRand m_rngRange; // Noise simulator for the range measurement
+  vpGaussRand m_rngElevAngle; // Noise simulator for the elevation angle measurement
+};
+
+/**
+ * \brief Class to simulate a flying aircraft.
+ */
+class  vpACSimulator
```
Comment (fspindle) [inline tutorial/kalman/tutorial-ukf-nonlinear-example.cpp:175]: Remove extra space `class vpACSimulator`

File: doc/tutorial/python/tutorial-install-python-bindings.dox (hunk 1 above)
Comment (rolalaro) [reply doc/tutorial/python/tutorial-install-python-bindings.dox:682]: Done in 407c96aa0b683ac4183b234ca0e5d6000ef4d22c

File: doc/tutorial/python/tutorial-install-python-bindings.dox (hunk 2 above)
Comment (rolalaro) [reply doc/tutorial/python/tutorial-install-python-bindings.dox:688]: Done in 407c96aa0b683ac4183b234ca0e5d6000ef4d22c

File: doc/tutorial/python/tutorial-install-python-bindings.dox (hunk 3 above)
Comment (rolalaro) [reply doc/tutorial/python/tutorial-install-python-bindings.dox:712]: Done in 407c96aa0b683ac4183b234ca0e5d6000ef4d22c

File: modules/core/include/visp3/core/vpMatrix.h (hunk 4 above)
Comment (rolalaro) [reply modules/core/include/visp3/core/vpMatrix.h:358]: Done in 8dbc865630951c165d3c5cdbecda6d69bfdbda4e

File: modules/core/src/math/matrix/vpMatrix.cpp (hunk 5 above)
Comment (rolalaro) [reply modules/core/src/math/matrix/vpMatrix.cpp:131]: Each time I save after correcting the missing space, it is erased by VSCode. There must be a problem in the configuration again.

File: tutorial/kalman/tutorial-ukf-nonlinear-example.cpp (hunk 17 above)
Comment (rolalaro) [reply tutorial/kalman/tutorial-ukf-nonlinear-example.cpp]: Done in 7b87be38904131169258f4a897112fed9c0c43e6

File: tutorial/kalman/tutorial-ukf-nonlinear-complex-example.cpp (hunk 16 above)
Comment (rolalaro) [reply tutorial/kalman/tutorial-ukf-nonlinear-complex-example.cpp]: Done

File: tutorial/kalman/tutorial-ukf-linear-example.cpp (hunk 14 above)
Comment (rolalaro) [reply tutorial/kalman/tutorial-ukf-linear-example.cpp]: Done in 75aedad45a73e122e6238441cfa956e3caa4bd16

File: modules/core/src/math/matrix/vpMatrix.cpp (hunk 6 above)
Comment (rolalaro) [reply modules/core/src/math/matrix/vpMatrix.cpp:5013]: Corrected in a9e22bbee

File: modules/core/src/math/matrix/vpMatrix.cpp (hunk 7 above)
Comment (rolalaro) [reply modules/core/src/math/matrix/vpMatrix.cpp:6092]: Corrected in a9e22bbee

File: modules/core/src/math/matrix/vpMatrix.cpp (hunk 8 above)
Comment (rolalaro) [reply modules/core/src/math/matrix/vpMatrix.cpp:6160]: Corrected in a9e22bbee

File: modules/core/src/math/matrix/vpMatrix.cpp (hunk 9 above)
Comment (rolalaro) [reply modules/core/src/math/matrix/vpMatrix.cpp:6237]: Corrected in a9e22bbee

File: modules/core/src/math/matrix/vpMatrix.cpp (hunk 10 above)
Comment (rolalaro) [reply modules/core/src/math/matrix/vpMatrix.cpp:6478]: Corrected in a9e22bbee

File: modules/core/include/visp3/core/vpUnscentedKalman.h [hunk 19]
```diff
@@ -0,0 +1,430 @@
...
+  Be \f$ \textbf{x} \in \mathcal{R}^n \f$ the internal state of the UKF and \f$ \textbf{P} \f$ the process covariance matrix.
+  We have:
+
+  \f{eqnarray*}{
+      \chi &=& sigma-function(\textbf{x}, \textbf{P}) \\
+      \textbf{w}^m, \textbf{w}^c &=& weight-function(n, parameters)
+   \f}
+
+  There are different ways of drawing the sigma points and associated weights in the litterature, such as the one
+  proposed by Julier or the one proposed by E. A. Wan and R. van der Merwe.
```
Comment (s-trinh) [inline modules/core/include/visp3/core/vpUnscentedKalman.h:70]: I would suggest to add Bibtex entries for these two references.

File: modules/core/src/math/kalman/vpUKSigmaDrawerMerwe.cpp [hunk 20]
```diff
@@ -0,0 +1,92 @@
...
+ * WARRANTY OF DESIGN, MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE.
+ *
+ * Description:
+ * Kalman filtering.
+ *
+*****************************************************************************/
+
+/*!
+  \file vpUKSigmaDrawerMerwe.cpp
+  \brief Sigma points drawer following the E. A. Wan and R. van der Merwe's method.
```
Comment (s-trinh) [inline modules/core/src/math/kalman/vpUKSigmaDrawerMerwe.cpp:38]: Same.
